### PR TITLE
improve types dx

### PR DESCRIPTION
**Pull request by zaceno**
Improves the DX of the types without relaxing any constraints. (As far as I can tell)

First: we distinguish between ElementVNodes and TextVNodes. That is important because ElementVNodes should be generic – parameterised on the state, so that various actions in a tree all work on the same state. This does not apply to text-nodes which a) have no actions assigned to them, and b) are always leaf-nodes (no children). For that reason, it doesn't make sense for text-nodes to be parameterized on the state, and also it doesn't make sense for the `text()` function to be a generic function.

Second: Whenever there are nodes, or parts of the tree that don't contain any actions, that part of the view tree will end up having the type `ElementVNode<unknown>` (prev `VNode<unknonw>`).
If such a node has siblings that are `VNode<SomeKNownState>` there will be a type error unless you explicitly pass the SomeKnownState as a generic to the `h` functions. This get's troublesome fast – especially when you are using JSDoc to type reusable components (which should not be aware of the state). 

How I fixed that was simply to allow: `MaybeVNode<S> = boolean | VNode<S> | ElementVNode<unknown>` As far as I can see that fixed all inference trouble for me, without relaxing anything (the State is still properly inferred and doesent "bail out" to any or unknown)

### CODE REVIEW

**Comment by zaceno** [comment]
Sorry it seems prettier did a number on the diff. Will clean it up shortly

**Review by icylace — CHANGES_REQUESTED** [review]
@zaceno I like the idea of this refactor. Fan fact: Relieving `text()` of a state type avoids a couple unsolved edge cases in my types test suite !

File: index.d.ts [hunk 1]
```diff
@@ -134,7 +134,12 @@ declare module "hyperapp" {
   }
 
   // In certain places a virtual DOM node can be made optional.
-  type MaybeVNode<S> = boolean | null | undefined | VNode<S>
+  type MaybeVNode<S> =
+    | boolean
+    | null
+    | undefined
+    | VNode<S>
+    | ElementVNode<unknown>
```
[the remark this answers is not in the document]
Comment (zaceno) [reply index.d.ts]: Will do! After having experimented some with this I agree it is unnecessary, and I also realized it does cause some loosening in some cases so it's gone!

File: index.d.ts [hunk 2]
```diff
@@ -47,10 +47,10 @@ declare module "hyperapp" {
   ): VNode<S>
```
[the remark this answers is not in the document]
Comment (zaceno) [reply index.d.ts]: 👍

**Comment by zaceno** [comment]
@icylace Thanks for the review – very good points, all! I have adressed them now, so I think we should be good to merge here. (Feel free to have another look and make sure!)

**Comment by icylace** [comment]
@zaceno @jorgebucaran Looks good to me !

**Comment by zaceno** [comment]
Ok reverted the memo return type back to VNode<S> 😅 !

**Comment by icylace** [comment]
@zaceno @jorgebucaran Okay, now I think we're good !
